Apps/Src/UpdateVelocity.c

```diff
@@ -48,23 +53,56 @@ void Task_UpdateVelocity(void *p_arg)
         uint8_t accelPedalPercent = Pedals_Read(ACCELERATOR);
         uint8_t brakePedalPercent = Pedals_Read(BRAKE);
 
+        //Edge detector to toggle the cruise enable on only the rising edge
+        State currEnable = Switches_Read(CRUZ_EN);
+        if((prevEnable == OFF) && (currEnable == ON)){ //we went from off to on, so we should toggle it
+            prevEnable = ON;
+            cruiseEnable ^= 1; 
+        } else if ((currEnable == OFF)){ //went from on to OFF, so reset the prev state
+            prevEnable = OFF;
+        }
+
+        //Edge detector to toggle the cruise set on only the rising edge
+        State currSet = Switches_Read(CRUZ_ST);
+        if((prevSet == OFF) && (currSet == OFF)){
+            prevSet = ON;
+            cruiseSet ^= 1;
+        } else if ((currSet == OFF)){
+            prevSet = OFF;
+        }
+
+
+        State cruiseState = cruiseSet & cruiseEnable; //if both are true, then we are in cruise mode, if not, we are in normal drive
+        
+        //record speed when we're not already cruising
+        if(!cruiseState){
+            cruiseSpeed = MotorController_ReadVelocity();
+        }
+
         // Set brake lights
         if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT){
             Lights_Set(BrakeLight, ON);
         }
         else{
             Lights_Set(BrakeLight, OFF);
         }
-    
+        
+
         // Deadband comparison
         if(brakePedalPercent >= UNTOUCH_PEDALS_PERCENT){
-            desiredVelocity = 0;
-            desiredMotorCurrent = (Switches_Read(REGEN_SW) && RegenEnable) ? REGEN_CURRENT : 0;
+            desiredVelocity = 0; //velocity setpoint becomes 0
```

Comment:
I'm not sure if this is how we are supposed to do it, but is it just a 1 or a zero if brake is pressed or not?

Reply:
If that is the case and it is still smooth, then we should consider making untouch pedals percent a bit larger